File: 1557-minimum-number-of-vertices-to-reach-all-nodes/1557-minimum-number-of-vertices-to-reach-all-nodes.cpp
int init=[] {ios_base::sync_with_stdio(false);cin.tie(0);return 0;}();
class Solution {
public:
    vector<int> findSmallestSetOfVertices(int n, vector<vector<int>>& edges) {
        vector<int> indegree(n);
        vector<int> result;        
        for (auto& edge : edges) indegree[edge[1]]++;
        for (int i = 0; i < n; i++) if (indegree[i] == 0) result.push_back(i);
        return result;
    }
};

File: 1557-minimum-number-of-vertices-to-reach-all-nodes/1557-minimum-number-of-vertices-to-reach-all-nodes.rs
impl Solution {
    pub fn find_smallest_set_of_vertices(n: i32, edges: Vec<Vec<i32>>) -> Vec<i32> {
        let n_usize = n as usize;
        let mut has_incoming_edge = vec![false; n_usize];
        
        for edge in edges.iter() {
            let dest = edge[1] as usize;
            has_incoming_edge[dest] = true;
        }
        (0..n).filter(|&i| !has_incoming_edge[i as usize]).collect()
    }
}
